From a2f1b67c8f45dbe7a7769a1c9bf1f2c73bb48d37 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Tue, 26 Aug 2014 12:49:33 -0400
Subject: [tools] cpupower: If root, try to load msr driver on x86 if /dev/cpu/0/msr is not available

Message-id: <1409057382-26432-23-git-send-email-prarit@redhat.com>
Patchwork-id: 89107
O-Subject: [RHEL7.1 PATCH BZ 1133893 22/31] cpupower: If root, try to load msr driver on x86 if /dev/cpu/0/msr is not available
Bugzilla: 1133893
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1133893

commit 8a19cb586708361058e089b7c23b6f3eb33af6c6
Author: Thomas Renninger <trenn@suse.de>
Date:   Tue May 13 12:41:43 2014 +0200

    cpupower: If root, try to load msr driver on x86 if /dev/cpu/0/msr is not available

    Signed-off-by: Thomas Renninger <trenn@suse.de>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Tony Camuso <tcamuso@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/power/cpupower/utils/cpupower.c b/tools/power/cpupower/utils/cpupower.c
index 7efc570..7cdcf88 100644
--- a/tools/power/cpupower/utils/cpupower.c
+++ b/tools/power/cpupower/utils/cpupower.c
@@ -12,6 +12,9 @@
 #include <string.h>
 #include <unistd.h>
 #include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/utsname.h>
 
 #include "builtin.h"
 #include "helpers/helpers.h"
@@ -169,6 +172,8 @@ int main(int argc, const char *argv[])
 {
  const char *cmd;
  unsigned int i, ret;
+ struct stat statbuf;
+ struct utsname uts;
 
  cpus_chosen = bitmask_alloc(sysconf(_SC_NPROCESSORS_CONF));
 
@@ -195,6 +200,15 @@ int main(int argc, const char *argv[])
 
  get_cpu_info(0, &cpupower_cpu_info);
  run_as_root = !getuid();
+ if (run_as_root) {
+  ret = uname(&uts);
+  if (!ret && !strcmp(uts.machine, "x86_64") &&
+      stat("/dev/cpu/0/msr", &statbuf) != 0) {
+   if (system("modprobe msr") == -1)
+ fprintf(stderr, _("MSR access not available.\n"));
+  }
+ }
+  
 
  for (i = 0; i < ARRAY_SIZE(commands); i++) {
   struct cmd_struct *p = commands + i;
-- 
1.7.1